### PR TITLE
🏗 Refactor Architecture Step 1

### PR DESCRIPTION
**Pull request by Vyperus**
**Changes:**

1. Delete `processdef-detail` entirely.
2. Adjust route parameters.
3. Add new Service: `SolutionService` which holds all currently open solutions.
4. Make all components use this service.
5. Refactor Inspect & design components.
6. Delete unused/unnecessary events.

PR: #1148 

## How can others test the changes?

Test this by running the full testing script normally used for releases.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

**Comment by cmg-dev** [comment]
@Vyperus I adjusted the language on your PR. Please use an active phrasing over a passive one.

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 1]
```diff
@@ -4,71 +4,89 @@ import {Redirect, Router} from 'aurelia-router';
...
+
+    this._activeSolutionEntry = await this._solutionService.getActiveSolutionEntry();
+
+    const diagramNameIsSet: boolean = routeParameters.diagramName !== undefined;
+
+    if (diagramNameIsSet) {
+      this.activeDiagram = await this._activeSolutionEntry.service.loadDiagram(routeParameters.diagramName);
+    } else {
+      return;
+    }
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: Simplify this:

```ts
const diagramNameIsNotSet: boolean = routeParameters.diagramName === undefined;
if (diagramNameIsNotSet) {
  return;
}

this.activeDiagram = await this._activeSolutionEntry.service.loadDiagram(routeParameters.diagramName);
```

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 2]
```diff
@@ -261,30 +319,38 @@ export class DiagramDetail {
...
+   * If there is only one StartEvent this method will select this StartEvent by
+   * default.
    */
-  private async _saveDiagram(): Promise<void> {
+  public async showSelectStartEventDialog(): Promise<void> {
+    await this._updateProcessStartEvents();
 
-    if (this._diagramIsInvalid) {
-      this._notificationService.showNotification(NotificationType.WARNING, `The could not be saved because it is invalid!`);
+    if (this.processesStartEvents.length === 1) {
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: Extract Business Rule from `if`

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 3]
```diff
@@ -332,4 +404,69 @@ export class DiagramDetail {
       .publish(environment.events.navBar.noValidationError);
     this._diagramIsInvalid = false;
   }
+
+    /**
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: correct indent level

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 4]
```diff
@@ -332,4 +404,69 @@ export class DiagramDetail {
...
+   * in the Property Panel, also split this into two methods and name them right.
+   */
+  private _dropInvalidFormData(): void {
+    const registry: IElementRegistry = this.bpmnio.modeler.get('elementRegistry');
+
+    registry.forEach((element: IShape) => {
+      if (element.type === 'bpmn:UserTask') {
+        const businessObj: IModdleElement = element.businessObject;
+
+        if (businessObj.extensionElements) {
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: Extract Business Rule from `if`

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 5]
```diff
@@ -332,4 +404,69 @@ export class DiagramDetail {
...
+   * FormData to make sure the diagrams XML is further supported by Camunda.
+   *
+   * TODO: Look further into this if this method is not better placed at the FormsSection
+   * in the Property Panel, also split this into two methods and name them right.
+   */
+  private _dropInvalidFormData(): void {
+    const registry: IElementRegistry = this.bpmnio.modeler.get('elementRegistry');
+
+    registry.forEach((element: IShape) => {
+      if (element.type === 'bpmn:UserTask') {
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: Extract Business Rule from `if`

```suggestion
      const elementTypeIsUserTask = element.type === 'bpmn:UserTask';
      if (elementTypeIsUserTask) {
```

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 6]
```diff
@@ -332,4 +404,69 @@ export class DiagramDetail {
...
+
+          extensions.values = extensions.values.filter((value: IFormElement) => {
+            const typeIsNotCamundaFormData: boolean = value.$type !== 'camunda:FormData';
+            const elementContainsFields: boolean = (value.fields !== undefined) && (value.fields.length > 0);
+
+            const keepThisValue: boolean = typeIsNotCamundaFormData || elementContainsFields;
+            return keepThisValue;
+          });
+
+          if (extensions.values.length === 0) {
```
Comment (cmg-dev) [inline src/modules/design/diagram-detail/diagram-detail.ts]: Extract Business Rule from `if`

File: src/modules/design/property-panel/indextabs/general/service/general.service.ts [hunk 7]
```diff
@@ -29,8 +29,8 @@ export class GeneralService {
     return this._generalRepository.getAllDiagrams();
   }
 
-  public updateDiagram(diagram: IDiagram): Promise<IDiagram> {
-    return this._generalRepository.updateDiagram(diagram);
-  }
+  // public updateDiagram(diagram: IDiagram): Promise<IDiagram> {
+  //   return this._generalRepository.updateDiagram(diagram);
+  // }
```
Comment (cmg-dev) [inline src/modules/design/property-panel/indextabs/general/service/general.service.ts]: if we do not need it -> delete it.

File: src/modules/solution-explorer-services/RefreshingSolutionExplorerService.ts [hunk 8]
```diff
@@ -61,13 +61,13 @@ export class RefreshingSolutionExplorerService implements ISolutionExplorerServi
...
+  // public openSingleDiagram(pathToDiagram: string, identity: IIdentity): Promise<IDiagram> {
+  //   return this._parent.openSingleDiagram(pathToDiagram, identity);
+  // }
 
-  public saveSingleDiagram(diagramToSave: IDiagram, identity: IIdentity, path?: string): Promise<IDiagram> {
-    return this._parent.saveSingleDiagram(diagramToSave, identity, path);
-  }
+  // public saveSingleDiagram(diagramToSave: IDiagram, identity: IIdentity, path?: string): Promise<IDiagram> {
+  //   return this._parent.saveSingleDiagram(diagramToSave, identity, path);
+  // }
```
Comment (cmg-dev) [inline src/modules/solution-explorer-services/RefreshingSolutionExplorerService.ts]: @Paulomart what were they for?

File: src/modules/solution-explorer-services/SingleDiagramsSolutionExplorerService.ts [hunk 9]
```diff
@@ -71,13 +72,16 @@ export class SingleDiagramsSolutionExplorerService implements ISolutionExplorerS
...
 
     if (uriAlreadyOpened) {
       throw new Error('This diagram is already opened.');
     }
 
+    await this._solutionExplorerToOpenDiagrams.openSolution(uri.substring(0, uri.lastIndexOf('/')), identity);
+
     const diagram: IDiagram = await this._solutionExplorerToOpenDiagrams
-      .openSingleDiagram(uri, identity);
+      .loadDiagram(uri.substring(uri.lastIndexOf('/') + 1, uri.length - bpmnFileEndingLength));
```
Comment (cmg-dev) [inline src/modules/solution-explorer-services/SingleDiagramsSolutionExplorerService.ts]: `uri.substring(uri.lastIndexOf('/') + 1, uri.length - bpmnFileEndingLength)`

extract this... give it a good name... tomorrow everyone forgets what this has done.

File: src/services/ActiveSolutionDiagramService.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import {IDiagram} from '@process-engine/solutionexplorer.contracts';
+import {ISolutionEntry, ISolutionService} from '../contracts';
+
+export class SolutionService implements ISolutionService {
+  private _allSolutionEntries: Array<ISolutionEntry> = [];
+  private _activeSolution: ISolutionEntry;
```
Comment (cmg-dev) [inline src/services/ActiveSolutionDiagramService.ts]:
```suggestion
  private _activeSolutionReference: ISolutionEntry;
```

As this should be no copy of the active solution, but instead the reference to one in the list!

File: src/services/ActiveSolutionDiagramService.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+  public getSolutionEntryForUri(uri: string): ISolutionEntry {
+    const solutionEntry: ISolutionEntry = this._allSolutionEntries.find((entry: ISolutionEntry) => {
+      return entry.uri === uri;
+    });
+
+    return solutionEntry;
+  }
+
+  public setActiveSolution(solution: ISolutionEntry): void {
```
Comment (cmg-dev) [inline src/services/ActiveSolutionDiagramService.ts]: does this need to be public?

File: src/contracts/solution-explorer/ISolutionService.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+   */
+  closeSolutionEntry(solutionEntry: ISolutionEntry): void;
+
+  /**
+   * Gets the currently active SolutionEntry.
+   */
+  getActiveSolutionEntry(): ISolutionEntry;
+
+  /**
+   * Gets a specific SolutionEntry identified by its uri.
```
Comment (cmg-dev) [inline src/contracts/solution-explorer/ISolutionService.ts]:
```suggestion
   * Gets a specific SolutionEntry identified by its URI.
```

File: src/modules/inspect/process-list/process-list.ts [hunk 13]
```diff
@@ -5,6 +5,7 @@ import {Router} from 'aurelia-router';
 import {IIdentity} from '@essential-projects/iam_contracts';
 import {Correlation, CorrelationProcessModel, IManagementApi} from '@process-engine/management_api_contracts';
 
+import { IDiagram } from '@process-engine/solutionexplorer.contracts';
```
Comment (alexanderkasten) [inline src/modules/inspect/process-list/process-list.ts]: Remove whitspaces

File: src/modules/navbar/navbar.ts [hunk 14]
```diff
@@ -208,6 +108,14 @@ export class NavBar {
         this.disableDashboardButton = false;
         this.disableInspectCorrelationButton = true;
       }),
+
+      this._eventAggregator.subscribe(environment.events.navBar.updateActiveSolutionAndDiagram, ({solutionEntry, diagram}: any) => {
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]:
```suggestion
      this._eventAggregator.subscribe(environment.events.navBar.updateActiveSolutionAndDiagram,
         ({solutionEntry, diagram}: {solutionEntry: ISolutionEntry, diagram: IDiagram}) => {
```

File: src/modules/navbar/navbar.ts [hunk 15]
```diff
@@ -288,44 +196,36 @@ export class NavBar {
...
-    const processModelId: string = (this.process && !activeRouteIsNotProcessEngineRoute)
-                                  ? this.process.id
-                                  : undefined;
+    let diagramName: string;
+    const diagramIsNotSelect: boolean = this.activeDiagram === undefined;
 
-    if (activeRouteIsNotProcessEngineRoute) {
-      this.showInspectTools = false;
-      this.showProcessName = false;
+    if (diagramIsNotSelect) {
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]:
```suggestion
    diagramIsNotSelect ? diagramName = undefined : diagramName = this.activeDiagram.name;
```

File: src/modules/navbar/navbar.ts [hunk 16]
```diff
@@ -243,7 +151,7 @@ export class NavBar {
     this.disableInspectCorrelationButton = false;
 
     this._router.navigateToRoute('inspect', {
-      processModelId: this.process.id,
+      diagramName: this.activeDiagram.name,
       view: 'dashboard',
       latestSource: this.latestSource,
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]: latestSoource was removed from Inspect.ts

File: src/modules/navbar/navbar.ts [hunk 17]
```diff
@@ -288,44 +196,36 @@ export class NavBar {
...
+    } else {
+      diagramName = this.activeDiagram.name;
     }
 
     this._router.navigateToRoute('inspect', {
-      processModelId: processModelId,
+      diagramName: diagramName,
       view: this.inspectView,
-      latestSource: undefined,
+      latestSource: this.latestSource,
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]: latestSource was removed from inspect.ts 🤔

File: src/modules/navbar/navbar.ts [hunk 18]
```diff
@@ -394,24 +294,26 @@ export class NavBar {
...
-      }
-    })();
+    const activeSolutionIsRemoteSolution: boolean = this.activeSolutionEntry.uri.startsWith('http');
+    this.showProcessName = this.activeDiagram.name !== undefined;
 
-    const latestSourceIsProcessEngine: boolean = this.latestSource === 'process-engine';
     this.navbarTitle = ((): string => {
-      if (latestSourceIsProcessEngine) {
-        return this.process.id;
+      if (activeSolutionIsRemoteSolution) {
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]:
```suggestion
      activeSolutionIsRemoteSolution ? return this.activeDiagram.id : return this.activeDiagram.name;
```

File: src/modules/inspect/inspect-correlation/components/diagram-viewer/diagram-viewer.ts [hunk 19]
```diff
@@ -119,7 +120,7 @@ export class DiagramViewer {
     await this._importXml(this._diagramViewer, this._uncoloredXml);
   }
 
-  public processModelIdChanged(): void {
+  public activeDiagramChanged(): void {
     if (this._diagramViewer === undefined) {
```
Comment (SteffenKn) [inline src/modules/inspect/inspect-correlation/components/diagram-viewer/diagram-viewer.ts]: Extract Business Rule, when already touching this file you could do this as well.

```suggestion
    const diagramViewerNotSet: boolean = this._diagramViewer === undefined;
    if (diagramViewerNotSet) {
```

File: src/modules/inspect/process-list/process-list.ts (hunk 13 above)
Comment (SteffenKn) [inline src/modules/inspect/process-list/process-list.ts]: Remove unneeded whitespaces
```suggestion
import {IDiagram} from '@process-engine/solutionexplorer.contracts';
```

File: src/modules/navbar/navbar.ts [hunk 20]
```diff
@@ -288,44 +196,36 @@ export class NavBar {
       return;
     }
 
-    const activeRouteIsNotProcessEngineRoute: boolean = this.activeRouteName !== 'processdef-detail';
-
-    const processModelId: string = (this.process && !activeRouteIsNotProcessEngineRoute)
-                                  ? this.process.id
-                                  : undefined;
+    let diagramName: string;
```
Comment (SteffenKn) [inline src/modules/navbar/navbar.ts]: You could set this via:
```
    const diagramIsNotSelect: boolean = this.activeDiagram === undefined;
    const diagramName: string = diagramIsNotSelected
      ? undefined
      : this.activeDiagram.name;
	
```

File: src/modules/solution-explorer/solution-explorer-list/solution-explorer-list.ts [hunk 21]
```diff
@@ -197,7 +196,8 @@ export class SolutionExplorerList {
         nameOfSingleDiagramService,
       );
 
-    this._addSolutionEntry(uriOfSingleDiagramService, this._singleDiagramService, true);
+    const identity: IIdentity = this._createIdentityForSolutionExplorer();
+    this._addSolutionEntry(uriOfSingleDiagramService, this._singleDiagramService, identity, true);
```
Comment (SteffenKn) [inline src/modules/solution-explorer/solution-explorer-list/solution-explorer-list.ts]: Add some spacing
```suggestion

    this._addSolutionEntry(uriOfSingleDiagramService, this._singleDiagramService, identity, true);
```

File: src/main.ts [hunk 22]
```diff
@@ -39,6 +39,8 @@ export function configure(aurelia: Aurelia): void {
   ? customProcessEngineRoute
   : window.localStorage.getItem('InternalProcessEngineRoute');
 
+  window.localStorage.setItem('processEngineRoute', processEngineRoute);
```
Comment (cmg-dev) [inline src/main.ts]: 👍

Comment (cmg-dev) [reply src/main.ts]: wir müssen das aber irgendwie in einen eigenen Service packen, sowas würde unter iOS nicht laufen.

File: src/modules/navbar/navbar.ts [hunk 23]
```diff
@@ -394,24 +294,24 @@ export class NavBar {
...
-    this.processOpenedFromProcessEngine = latestSourceIsProcessEngine;
+    this.processOpenedFromProcessEngine = activeSolutionIsRemoteSolution;
   }
+
+  private _updateNavbarTools(): void {
+    const activeSolutionIsRemoteSolution: boolean = this.activeSolutionEntry.uri.startsWith('http');
+
+    this.disableStartButton = !activeSolutionIsRemoteSolution;
+    this.disableDiagramUploadButton = activeSolutionIsRemoteSolution;
+    this.disableDiagramUploadButton = activeSolutionIsRemoteSolution;
```
Comment (alexanderkasten) [inline src/modules/navbar/navbar.ts]: why is this double?

**Comment by robpal94** [comment]
After the Process Execution is finished, the following error is thrown: 
```
scripts/vendor-bundle.js:14870 ERROR [app-router] NotFoundError: ProcessModel with id undefined not found!
    at HttpFetchClient.<anonymous> (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/app-bundle.js:2878:39)
    at step (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:29129:23)
    at Object.next (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:29110:53)
    at fulfilled (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:29101:58)
    at tryCatcher (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:3434:23)
    at Promise._settlePromiseFromHandler (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:2649:31)
    at Promise._settlePromise (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:2706:18)
    at Promise._settlePromise0 (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:2751:10)
    at Promise._settlePromises (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:2831:18)
    at _drainQueueStep (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:168:12)
    at _drainQueue (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:161:9)
    at Async._drainQueues (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:177:5)
    at Async.drainQueues (file:///Users/robin/Documents/process_engine/bpmn-studio/scripts/vendor-bundle.js:47:14)
    at <anonymous>
```

screenshot 2018-11-30 at 15 35 42

File: src/main.ts (hunk 22 above)
Comment (cmg-dev) [reply src/main.ts]: genau das wollte ich hören.

**Pull request merged**